shared/health/src/main/java/io/pravega/shared/health/impl/ContributorRegistryImpl.java

```diff
@@ -0,0 +1,312 @@
+                unregister(child);
+            }
+        }
+        // Remove contributor reference from all containers.
+        parents.remove(name);
+        children.remove(name);
+        contributors.remove(name);
+
+        return contributor;
+    }
+
+    /**
+     * This method returns an {@link Optional} because it signifies that one should account for the case that
+     * there may be no {@link String} to {@link HealthContributor} mapping.
+     *
+     * @param name The name of some {@link HealthContributor} to request from this {@link ContributorRegistry}.
+     * @return The {@link HealthContributor} mapped to by 'name'.
+     */
+    @Override
+    @NonNull
+    public Optional<HealthContributor> get(String name) {
+        return Optional.ofNullable(contributors.get(name));
+    }
+
+    /**
+     * Provides a reference to the root {@link HealthContributor}.
+     * @return The root {@link HealthContributor}.
+     */
+    @Override
+    public HealthContributor get() {
+        return contributors.get(DEFAULT_CONTRIBUTOR_NAME);
+    }
+
+    /**
+     * Provides the set of dependencies that the {@link HealthContributor} mapped to by 'name' relies on.
+     *
+     * @param name The {@link String} used to query which {@link HealthContributor} to gather it's {@link HealthContributor}
+     *             dependencies.
+     * @return The {@link Collection} used to determine the {@link io.pravega.shared.health.Health}
+     *         of the {@link HealthContributor} mapped to by 'name'.
+     */
+    @Override
+    public Collection<HealthContributor> dependencies(String name) {
+        return this.children.get(name);
+    }
+
+    /**
+     *
+     * Provides the set of dependencies that the root {@link HealthContributor} relies on.
+     *
+     * @return The {@link Collection} of {@link HealthContributor} used by the root.
+     */
+    @Override
+    public Collection<HealthContributor> dependencies() {
+        return dependencies(DEFAULT_CONTRIBUTOR_NAME);
+    }
+
+    /**
+     * Supplies a {@link Collection} of *all* {@link HealthContributor} ids (names) registered within
+     * this {@link ContributorRegistry}.
+     *
+     * @return The {@link Collection} of {@link HealthContributor} names.
+     */
+    @Override
+    public Collection<String> contributors() {
+        return contributors.entrySet()
+                .stream()
+                .map(entry -> entry.getKey())
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Supplies a {@link Collection} of *all* {@link HealthComponent} ids (names) registered within
+     * this {@link ContributorRegistry}. This method returns a subset of the {@link Collection} returned by
+     * {@link ContributorRegistry#contributors()}.
+     *
+     * @return The {@link Collection} of {@link HealthComponent} names.
+     */
+    public Collection<String> components() {
+        return this.components;
```

Comment:
You are giving away your internal collection. Your caller can now do anything they want with it, including modifying it. Is that what you want?